File: samples/authorization/app/src/main/java/de/telekom/scauthorizationdemo/mappers/ConfigurationMapper.java
package de.telekom.scauthorizationdemo.mappers;

import androidx.appcompat.app.AppCompatActivity;

import org.json.JSONException;

import de.telekom.scauthorizationdemo.DemoApplication;
import de.telekom.scauthorizationdemo.activities.ConfigurationActivity;
import de.telekom.scauthorizationdemo.callbacks.RetrieveConfigurationCallback;
import de.telekom.scauthorizationdemo.task.SmartTask;
import de.telekom.smartcredentials.core.api.StorageApi;
import de.telekom.smartcredentials.core.authorization.AuthorizationConfiguration;
import de.telekom.smartcredentials.core.authorization.AuthorizationView;
import de.telekom.smartcredentials.core.filter.SmartCredentialsFilterFactory;
import de.telekom.smartcredentials.core.itemdatamodel.ItemEnvelope;
import de.telekom.smartcredentials.core.responses.SmartCredentialsApiResponse;
import de.telekom.smartcredentials.storage.factory.SmartCredentialsStorageFactory;
import timber.log.Timber;


public class ConfigurationMapper {

    private final AppCompatActivity activity;
    private final StorageApi storageApi;

    public ConfigurationMapper(AppCompatActivity activity) {
        this.activity = activity;
        this.storageApi = SmartCredentialsStorageFactory.getStorageApi();
    }

    public void retrieveAuthorizationConfiguration(RetrieveConfigurationCallback callback) {
        SmartTask.with(activity)
                .assign(() -> storageApi.getItemDetailsById(SmartCredentialsFilterFactory.createNonSensitiveItemFilter(
                        ConfigurationActivity.ITEM_ID,
                        ConfigurationActivity.ITEM_TYPE)))
                .finish(result -> {
                    SmartCredentialsApiResponse<ItemEnvelope> response =
                            (SmartCredentialsApiResponse<ItemEnvelope>) result;
                    if (response != null && response.isSuccessful()) {
                        callback.onConfigurationRetrieved(generateConfiguration(response.getData()));
                    } else {
                        callback.onFailedToRetrieveConfiguration();
                        Timber.tag(DemoApplication.TAG).d("Failed to fetch item details.");
                    }
                })
                .execute();
    }

    private AuthorizationConfiguration generateConfiguration(ItemEnvelope itemEnvelope) {
        AuthorizationConfiguration authorizationConfiguration = null;
        try {
            AuthorizationView.Builder authorizationViewBuilder = new AuthorizationView.Builder(
                    itemEnvelope.getDetails().getString(ConfigurationActivity.ITEM_KEY_TITLE),
                    itemEnvelope.getDetails().getString(ConfigurationActivity.ITEM_KEY_NEGATIVE_BUTTON_TEXT));

            String subtitle = itemEnvelope.getDetails().getString(ConfigurationActivity.ITEM_KEY_SUBTITLE);
            if (!subtitle.isEmpty()) {
                authorizationViewBuilder.setSubtitle(subtitle);
            }

            String description = itemEnvelope.getDetails().getString(ConfigurationActivity.ITEM_KEY_DESCRIPTION);
            if (!description.isEmpty()) {
                authorizationViewBuilder.setDescription(description);
            }

            AuthorizationConfiguration.Builder authorizationConfigurationBuilder = new AuthorizationConfiguration
                    .Builder(authorizationViewBuilder.build());
            authorizationConfigurationBuilder.allowDeviceCredentialsFallback(
                    itemEnvelope.getDetails().getBoolean(ConfigurationActivity.ITEM_KEY_DEVICE_CREDENTIALS_FALLBACK));
            authorizationConfigurationBuilder.requireFaceRecognitionConfirmation(
                    itemEnvelope.getDetails().getBoolean(ConfigurationActivity.ITEM_KEY_REQUIRE_CONFIRMATION));
            authorizationConfiguration = authorizationConfigurationBuilder.build();
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return authorizationConfiguration;
    }
}

File: samples/qrlogin/app/src/main/java/de/telekom/scqrlogindemo/mapper/ConfigurationMapper.kt
package de.telekom.scqrlogindemo.mapper

import androidx.appcompat.app.AppCompatActivity
import de.telekom.scqrlogindemo.DemoApplication
import de.telekom.scqrlogindemo.callback.RetrieveConfigurationCallback
import de.telekom.scqrlogindemo.activity.UserConfigurationActivity
import de.telekom.scqrlogindemo.model.UserConfiguration
import de.telekom.scqrlogindemo.task.SmartTask
import de.telekom.smartcredentials.core.api.StorageApi
import de.telekom.smartcredentials.core.filter.SmartCredentialsFilterFactory
import de.telekom.smartcredentials.core.itemdatamodel.ItemEnvelope
import de.telekom.smartcredentials.core.responses.SmartCredentialsApiResponse
import de.telekom.smartcredentials.storage.factory.SmartCredentialsStorageFactory
import org.json.JSONException
import timber.log.Timber

class ConfigurationMapper(private val activity: AppCompatActivity) {

    val storageApi: StorageApi = SmartCredentialsStorageFactory.getStorageApi()

    fun retrieveUserConfiguration(callback: RetrieveConfigurationCallback) {
        SmartTask.with(activity)
            ?.assign(getRetrieveTaskDescription())
            ?.finish(getRetrieveFinishListener(callback))
            ?.execute()
    }

    private fun getRetrieveTaskDescription(): SmartTask.TaskDescription {
        return object : SmartTask.TaskDescription {
            override fun onBackground(): Any? {
                return storageApi.getItemDetailsById(
                    SmartCredentialsFilterFactory.createNonSensitiveItemFilter(
                        UserConfigurationActivity.ITEM_ID,
                        UserConfigurationActivity.ITEM_TYPE
                    )
                )
            }
        }
    }

    private fun getRetrieveFinishListener(callback: RetrieveConfigurationCallback): SmartTask.FinishListener {
        return object : SmartTask.FinishListener {
            override fun onFinish(result: Any?) {
                val response = result as SmartCredentialsApiResponse<ItemEnvelope>?
                if (response != null && response.isSuccessful) {
                    callback.onConfigurationRetrieved(generateUserConfiguration(response.data))
                } else {
                    callback.onFailedToRetrieveConfiguration()
                    Timber.tag(DemoApplication.TAG)
                        .d("Failed to fetch user configuration item details.")
                }
            }
        }
    }

    private fun generateUserConfiguration(itemEnvelope: ItemEnvelope?): UserConfiguration? {
        var userConfiguration: UserConfiguration? = null
        try {
            val itemDetails = itemEnvelope?.details
            userConfiguration = itemDetails?.let {
                UserConfiguration(
                    it.getString(UserConfigurationActivity.ITEM_KEY_USERNAME),
                    it.getString(UserConfigurationActivity.ITEM_KEY_PASSWORD),
                    it.getString(UserConfigurationActivity.ITEM_KEY_URL),
                    it.getString(UserConfigurationActivity.ITEM_KEY_WEB_SOCKET)
                )
            }
        } catch (e: JSONException) {
            e.printStackTrace()
        }
        return userConfiguration
    }
}